### PR TITLE
Live Updating Upcoming Departures

### PR DESCRIPTION
**Pull request by zovt**
#### Summary of changes
**Asana Ticket:** [⏱️Schedule Finder | Upcoming Departures not updating live](https://app.asana.com/0/555089885850811/1150037444841984/f)

This is a pretty chonky refactor of this component. Notable things include:
- new utility hooks / functions:
  - `useProvider`
  - `useAwaitInterval`
  - `fetchJson`
- utilizing the above ^ to simplify the existing data fetching in the schedule finder (i.e. removing the reducers, different states, etc.)
   - Also, the network fetching code now just raises an exception since it's only done in one place and the previous code didn't really do anything particularly smart with the errors anyways.
- Behavior changes:
  - Instead of lazily fetching trip info data for non-bus modes, all modes now fetch the trip info data up front
  - The departure information and trip details in the modal should continually update while the modal is open (but not when the modal is closed)
  - **ANYTHING ELSE IS A BUG** (I think 😄)

Please ask questions about things that don't make sense. I've tried to keep the changes clear, but I'm happy to clarify anything.

I've also set the refresh rate to be ~once per second, but I think this could probably be extended. Let me know what you think!

### CODE REVIEW

**Review by lund5000 — CHANGES_REQUESTED** [review]
Nick and I observed what seems like a bug on https://dev-green.mbtace.com/schedules/1/line:  more than one prediction is visible in the line diagram, but when the Schedule Finder is opened only one upcoming departure is listed. It also seems to be an issue in prod. 😬 We should fix it, but if it's too different from the current PR or messy we can split into a different task & PR.

Also before this is approved and merged it would be nice to get the LIVE icon back into the view now that we're actually updating live again: https://github.com/mbta/dotcom/pull/1151

**Comment by thecristen** [comment]
> Nick and I observed what seems like a bug on https://dev-green.mbtace.com/schedules/1/line: more than one prediction is visible in the line diagram, but when the Schedule Finder is opened only one upcoming departure is listed. It also seems to be an issue in prod. 😬 We should fix it, but if it's too different from the current PR or messy we can split into a different task & PR.

I didn't notice that in the schedules I was looking at, that's interesting! Let's address that separately.

**Comment by zovt** [comment]
> > Nick and I observed what seems like a bug on https://dev-green.mbtace.com/schedules/1/line: more than one prediction is visible in the line diagram, but when the Schedule Finder is opened only one upcoming departure is listed. It also seems to be an issue in prod. grimacing We should fix it, but if it's too different from the current PR or messy we can split into a different task & PR.
> 
> I didn't notice that in the schedules I was looking at, that's interesting! Let's address that separately.

I think this may be a temporary API blip or something like that. We don't render table rows for trips without predictions, and for this particular route this morning, 4 out of the 5 trips we received from the backend didn't have predictions for some reason. Either way, I don't think this bug was introduced by this work so it makes sense to address it separately :+1:

**Review by thecristen — COMMENTED** [review]
Still poking at this, but liking everything I see so far. 🤩 

I was wondering if you're familiar with React Query at all - I haven't used it myself but have seen it held in high regard, and a lot of this code strikes me as functionality provided there, from [loading indication](https://tanstack.com/query/v4/docs/guides/queries) to [dependent queries](https://tanstack.com/query/v4/docs/guides/dependent-queries) to [refresh interval options](https://tanstack.com/query/v4/docs/reference/useQuery) ([async example](https://tanstack.com/query/v4/docs/examples/react/auto-refetching?from=reactQueryV3&original=https://react-query-v3.tanstack.com/examples/auto-refetching)). You should check it out and see whether it'll help you out here!

Will follow up with more comment on the code provided here soon. :)

**Review by thecristen — CHANGES_REQUESTED** [review]
This is such a nicer implementation! And the live updates are working great. Just a couple minor questions along with my earlier commentary.

With respect to Steve's earlier request to re-add our prior "Live" icon -- [here's the PR that had removed it](https://github.com/mbta/dotcom/pull/1151/files) - you can probably get away with just reverting it.

The next thing I'd want to check is performance. Naturally by making this live again, we incur many more network requests (illustrated below).

image

Can Dotcom handle this many requests?

@zovt, I'd like you to try some load testing here. We've got some [load testing tooling and scripts set up under `apps/site/load_tests`](https://github.com/mbta/dotcom/blob/master/apps/site/load_tests/README.md#L1) using `Locust`, which can simulate thousands of users hitting a resource together. Let's add a script here to demonstrate that we can hit these `/schedules/finder_api` endpoints many more times without crashing the application.

**Comment by arkadyan** [comment]
Bigger picture question—and caveat up front that I haven't actually looked at these code changes yet, I just read the PR note, so apologies if I could have answered my own question if I was a better reader…

I'm curious about the impetus behind using polling. I guess I tend to think of polling as (often) an anti-pattern (or a term one degree less severe than that) in Phoenix given that Channels are one of Phoenix's most awesome features. Would channels work here? I.e. could the client subscribe to updates and when the data changes on the server the changes are pushed to the client immediately? I don't know this example very well, but just at a quick glance, it looks like the [Map](https://github.com/mbta/dotcom/blob/master/apps/site/assets/ts/schedule/components/Map.tsx#L35) is one place we already use this.

File: apps/site/assets/ts/helpers/fetch-json.ts
Comment (zovt) [reply apps/site/assets/ts/helpers/fetch-json.ts:21]: typically I prefer not throwing in "library" functions by default since thrown exceptions are not typechecked, and this makes it explicit that the caller needs to handle the error. Definitely see the point here that it's kind of annoying having to call if the caller wants to throw though :) 

how do you feel about adding another export to this file, something like `fetchJsonOrThrow`, which has this behavior by default?

File: apps/site/assets/ts/helpers/use-await-interval.ts
Comment (zovt) [reply apps/site/assets/ts/helpers/use-await-interval.ts:17]: Essentially, the function that gets called by `setTimeout` (in this case, `go`) calls `setTimeout` again (via `enqueue`) after `await`ing the function passed into this hook

**Comment by zovt** [comment]
> I'm curious about the impetus behind using polling. I guess I tend to think of polling as (often) an anti-pattern (or a term one degree less severe than that) in Phoenix given that Channels are one of Phoenix's most awesome features. Would channels work here? I.e. could the client subscribe to updates and when the data changes on the server the changes are pushed to the client immediately? I don't know this example very well, but just at a quick glance, it looks like the [Map](https://github.com/mbta/dotcom/blob/master/apps/site/assets/ts/schedule/components/Map.tsx#L35) is one place we already use this.

I was not aware of Phoenix channels :) do you have any thoughts on how this would work? How would the backend know when to "push"? Since there are many permutations of this data possible (e.g. each individual route, the user selecting the start/end station, etc), I wouldn't want to maintain open connections for polling each possible combination of these from the API on the backend. Otherwise, if we're just lazily initializing polling on the backend and forwarding the result of that polling to the frontend, that doesn't seem like a huge difference vs. just polling from the front end.

Definitely open to exploring this direction more though! (this is good to keep in mind for the upcoming predictions work as well)

**Comment by zovt** [comment]
> Can Dotcom handle this many requests?

@thecristen 

For context, I'm not fully aware of exactly how often this feature is used, or of any existing performance concerns. Also, the behavior you are seeing of the multiple network requests is currently what happens in production with bus routes (since we need to fetch the crowding information up front to render the icons), aside from the reloading.

That said, I would be pretty surprised if this had a noticeable performance impact, since:
1. I believe these endpoints are fairly performant, at least no more taxing than the existing line diagram stuff
2. We're only ever sending up to 6 requests / refresh interval / page, since the endpoint that returns the departures only returns a maximum of 5, and there can only be one instance of the modal on each page.

If we're concerned about users potentially accidentally DoSing us by having multiple pages open or something, that sounds like something we should fix globally :)

Before I merge this, I can also adjust the refresh interval to be something more like 5-10 seconds if that seems more reasonable.

I think if anything, this could be something that gets addressed when we about the prediction unification work as well.

Anyways, let me know if that eases any of your concerns. I personally think the risk of deploying this is low enough that we should just try it out and eyeball any existing hardware performance metrics to see if there's any noticeable change.

**Comment by thecristen** [comment]
Re: Channels: I think that's a great way to go ultimately! @zovt: to see how it works you can look at `apps/vehicles/lib/stream.ex` as an example. The [API handles streaming new vehicle info](https://www.mbta.com/developers/v3-api/streaming), and we handle processing that here. The web application can subscribe to this stream via Phoenix channel, and there's a Javascript mechanism to hook into that.

It's very neat and I think we should totally use it for predictions, probably at a later point. I leave it up to you whether you want to follow that thread now or later! :)

With respect to this PR in its current state, I will note we've gotten burned in the past by polling implementation attempts that had to be hastily reverted. 😉 I'm happy to take a deploy-and-see approach here but we still need the `Live` icon.

**Comment by arkadyan** [comment]
It's streaming all the way down!

**Comment by zovt** [comment]
Interesting, thanks for the information! I think for now I'll opt to stick with the polling since this PR is already quite large and will probably end up getting gobbled up in the predictions rework work. But I do like the idea of moving to the push-based channels as part of that work!

W/r/t the other outstanding feedback, I'm planning on getting that all set today before merging :) thanks!

**Comment by zovt** [comment]
@thecristen @lund5000 I think this should be good to go for another round of review

**Comment by thecristen** [comment]
My bad, looks like there was a lot more to the old "live" icon 🤦🏼‍♀️ I'll submit a fix in a moment.

File: apps/site/assets/ts/helpers/fetch-json.ts
Comment (thecristen) [reply apps/site/assets/ts/helpers/fetch-json.ts:21]: > how do you feel about adding another export to this file, something like fetchJsonOrThrow, which has this behavior by default?

I like this idea!

File: apps/site/assets/ts/helpers/__tests__/use-await-interval-test.ts
Comment (thecristen) [inline apps/site/assets/ts/helpers/__tests__/use-await-interval-test.ts:48]: Really cool demonstration, thanks! 😊

File: apps/site/assets/ts/schedule/components/schedule-finder/ScheduleModalContent.tsx
Comment (thecristen) [inline apps/site/assets/ts/schedule/components/schedule-finder/ScheduleModalContent.tsx:104]: Thanks! Not sure how many folks come to Dotcom for down-to-the-second updates on the ground but I'm sure this'll suffice.
I hadn't had time to dig this up earlier, but now looking at `Site.TransitNearMe.format_prediction_time`, which formats the output times, we say "arriving" when a bus is < 60 seconds or a subway is < 30 seconds away. Which is to say we'd need an interval sufficiently small to be informative in those cases... which is to say 10 seconds will probably be great middle ground between polling too eagerly and not eagerly enough. Thanks!

**Review by thecristen — APPROVED** [review]

File: apps/site/assets/ts/schedule/components/schedule-finder/ScheduleModalContent.tsx
Comment (lund5000) [reply apps/site/assets/ts/schedule/components/schedule-finder/ScheduleModalContent.tsx:104]: +1 to switching from every second to every 10 seconds seeming like a reasonable middle ground

**Review by lund5000 — APPROVED** [review]
Excellent use of the <blink> tag

**Pull request merged**
